### ISSUE TITLE
Help to speed up inference processing

### ISSUE THREAD

**Issue by cnxdeverloper** [lukas-blecher/LaTeX-OCR#29]
Hi authors.
Thank you so much for awesome project. It working very good. Currently i got issues about time consuming when I ran with multiple cropped image formulation ( about 50 images) it took about 9 s to run all images. I ran model by function call_model in pix2text.py and with GPU 2080TI. Do you have some ideal to speed up inference processing. Thank u so much

**Comment by lukas-blecher** [comment]
Hi,
as of right now `call_model` does not work for batches of images. But it is possible to rewrite it to take multiple images at once, which would speed up inference.

How do you call the function right now? Did you write your own small script?
Depending on what size are the images are you can also try and use the [`evaluate`](https://github.com/lukas-blecher/LaTeX-OCR/blob/50e382349de9fccebea4460ee5a4ff9b9f9a2538/eval.py#L33) function in `eval.py`. But it would need some rewriting as well.

**Comment by cnxdeverloper** [comment]
> Hi,
> as of right now `call_model` does not work for batches of images. But it is possible to rewrite it to take multiple images at once, which would speed up inference.
> 
> How do you call the function right now? Did you write your own small script?
> Depending on what size are the images are you can also try and use the [`evaluate`](https://github.com/lukas-blecher/LaTeX-OCR/blob/50e382349de9fccebea4460ee5a4ff9b9f9a2538/eval.py#L33) function in `eval.py`. But it would need some rewriting as well.

Thank you for quickly reply. I ran with for loop to call function call_model . I will try to run with batchsize Thank you for your support

**Comment by lukas-blecher** [comment]
I will close this issue then. Feel free to reopen if you have further questions

**Comment by cnxdeverloper** [comment]
After deeply investigate code. I got a problem. After process by resize image in function call_model the output images are not same size so i can't stack or cat to 1 Torch Tensor to run batch size. How do you fixed the problem.

**Comment by lukas-blecher** [comment]
Yes, that is an issue here. While you can pad the images of one batch to the same size, the performance will most likely suffer and it would probably take even longer.
So you will only be able to batch images of similar size. I tried it out, it does speed up the prediction by a little bit:
I've batched together all images of the same size after the resizing step (I think this one has to be done indiviudally, not sure how to handle different image sizes here)

```python
# LaTeX-OCR/batchprediction.py
import os
import torch
import numpy as np
from dataset.dataset import *
from pix2tex import *
from PIL import Image
import re
from tqdm.auto import tqdm
from munch import Munch

img_path = ...
imgs = os.listdir(img_path)
pils = [Image.open(os.path.join(img_path, p)) for p in imgs]
batchsize = 5


args = Munch({'config': 'settings/config.yaml', 'checkpoint': 'checkpoints/weights.pth', 'no_cuda': False, 'no_resize': False})

args, model, image_resizer, tokenizer = initialize(args)
encoder, decoder = model.encoder, model.decoder


def predict(im):
    with torch.no_grad():
        model.eval()
        device = args.device
        encoded = encoder(im.to(device))
        dec = model.decoder.generate(torch.LongTensor([args.bos_token]*len(encoded))[:, None].to(device), args.max_seq_len,
                                     eos_token=args.pad_token, context=encoded, temperature=args.get('temperature', .2))
        pred = token2str(dec, tokenizer)
    return [post_process(s) for s in pred]


def img_resize(img):
    img = minmax_size(pad(img), args.max_dimensions, args.min_dimensions)
    if image_resizer is not None and not args.no_resize:
        with torch.no_grad():
            input_image = pad(img).convert('RGB').copy()
            r, w = 1, img.size[0]
            for i in range(5):
                img = minmax_size(input_image.resize((w, int(
                    input_image.size[1]*r)), Image.BILINEAR if r > 1 else Image.LANCZOS), args.max_dimensions, args.min_dimensions)
                t = test_transform(image=np.array(pad(img).convert('RGB')))['image'][:1].unsqueeze(0)
                w = image_resizer(t.to(args.device)).argmax(-1).item()*32
                if (w/img.size[0] == 1):
                    break
                r *= w/img.size[0]
        return t.to(args.device)
    else:
        return test_transform(image=np.array(pad(img).convert('RGB')))['image'][:1].unsqueeze(0)


image_tensors = [img_resize(pil) for pil in pils]
shapes = np.array([list(t.shape) for t in image_tensors])

batches = []
indices = []
shape_nums = (shapes[:, [-2, -1]]//32*np.array([args.max_width//32, 1])).sum(1)
for shape in np.unique(shapes, axis=0):
    inds = np.where(shape_nums == (args.max_width//32*shape[2]+shape[3])//32)[0]
    #print(shape, inds, shape_nums[inds])
    indices.extend(list(inds))
    tensors = torch.cat([image_tensors[j] for j in inds])
    for i in range(0, len(tensors), batchsize):
        batches.append(tensors[i:i+batchsize])


outputs = []
for batch in tqdm(batches):
    outputs.extend(predict(batch))
outputs = [outputs[i] for i in np.argsort(indices)]

print('\n'.join(outputs))
```

**Comment by cnxdeverloper** [comment]
> Yes, that is an issue here. While you can pad the images of one batch to the same size, the performance will most likely suffer and it would probably take even longer.
> So you will only be able to batch images of similar size. I tried it out, it does speed up the prediction by a little bit:
> I've batched together all images of the same size after the resizing step (I think this one has to be done indiviudally, not sure how to handle different image sizes here)
> 
> ```python
> # LaTeX-OCR/batchprediction.py
> import os
> import torch
> import numpy as np
> from dataset.dataset import *
> from pix2tex import *
> from PIL import Image
> from tqdm.auto import tqdm
> from munch import Munch
> 
> img_path = ...
> imgs = os.listdir(img_path)
> pils = [Image.open(os.path.join(img_path, p)) for p in imgs]
> batchsize = 5
> 
> 
> args = Munch({'config': 'settings/config.yaml', 'checkpoint': 'checkpoints/weights.pth', 'no_cuda': False, 'no_resize': False})
> 
> args, model, image_resizer, tokenizer = initialize(args)
> encoder, decoder = model.encoder, model.decoder
> 
> 
> def predict(im):
>     with torch.no_grad():
>         model.eval()
>         device = args.device
>         encoded = encoder(im.to(device))
>         dec = model.decoder.generate(torch.LongTensor([args.bos_token]*len(encoded))[:, None].to(device), args.max_seq_len,
>                                      eos_token=args.pad_token, context=encoded, temperature=args.get('temperature', .2))
>         pred = token2str(dec, tokenizer)
>     return [post_process(s) for s in pred]
> 
> 
> def img_resize(img):
>     img = minmax_size(pad(img), args.max_dimensions, args.min_dimensions)
>     if image_resizer is not None and not args.no_resize:
>         with torch.no_grad():
>             input_image = pad(img).convert('RGB').copy()
>             r, w = 1, img.size[0]
>             for i in range(5):
>                 img = minmax_size(input_image.resize((w, int(
>                     input_image.size[1]*r)), Image.BILINEAR if r > 1 else Image.LANCZOS), args.max_dimensions, args.min_dimensions)
>                 t = test_transform(image=np.array(pad(img).convert('RGB')))['image'][:1].unsqueeze(0)
>                 w = image_resizer(t.to(args.device)).argmax(-1).item()*32
>                 if (w/img.size[0] == 1):
>                     break
>                 r *= w/img.size[0]
>         return t.to(args.device)
>     else:
>         return test_transform(image=np.array(pad(img).convert('RGB')))['image'][:1].unsqueeze(0)
> 
> 
> image_tensors = [img_resize(pil) for pil in pils]
> shapes = np.array([list(t.shape) for t in image_tensors])
> 
> batches = []
> indices = []
> shape_nums = (shapes[:, [-2, -1]]//32*np.array([args.max_width//32, 1])).sum(1)
> for shape in np.unique(shapes, axis=0):
>     inds = np.where(shape_nums == (args.max_width//32*shape[2]+shape[3])//32)[0]
>     #print(shape, inds, shape_nums[inds])
>     indices.extend(list(inds))
>     tensors = torch.cat([image_tensors[j] for j in inds])
>     for i in range(0, len(tensors), batchsize):
>         batches.append(tensors[i:i+batchsize])
> 
> 
> outputs = []
> for batch in tqdm(batches):
>     outputs.extend(predict(batch))
> outputs = [outputs[i] for i in np.argsort(indices)]
> 
> print('\n'.join(outputs))
> `` i will try your code. Thanks

> Yes, that is an issue here. While you can pad the images of one batch to the same size, the performance will most likely suffer and it would probably take even longer.
> So you will only be able to batch images of similar size. I tried it out, it does speed up the prediction by a little bit:
> I've batched together all images of the same size after the resizing step (I think this one has to be done indiviudally, not sure how to handle different image sizes here)
> 
> ```python
> # LaTeX-OCR/batchprediction.py
> import os
> import torch
> import numpy as np
> from dataset.dataset import *
> from pix2tex import *
> from PIL import Image
> from tqdm.auto import tqdm
> from munch import Munch
> 
> img_path = ...
> imgs = os.listdir(img_path)
> pils = [Image.open(os.path.join(img_path, p)) for p in imgs]
> batchsize = 5
> 
> 
> args = Munch({'config': 'settings/config.yaml', 'checkpoint': 'checkpoints/weights.pth', 'no_cuda': False, 'no_resize': False})
> 
> args, model, image_resizer, tokenizer = initialize(args)
> encoder, decoder = model.encoder, model.decoder
> 
> 
> def predict(im):
>     with torch.no_grad():
>         model.eval()
>         device = args.device
>         encoded = encoder(im.to(device))
>         dec = model.decoder.generate(torch.LongTensor([args.bos_token]*len(encoded))[:, None].to(device), args.max_seq_len,
>                                      eos_token=args.pad_token, context=encoded, temperature=args.get('temperature', .2))
>         pred = token2str(dec, tokenizer)
>     return [post_process(s) for s in pred]
> 
> 
> def img_resize(img):
>     img = minmax_size(pad(img), args.max_dimensions, args.min_dimensions)
>     if image_resizer is not None and not args.no_resize:
>         with torch.no_grad():
>             input_image = pad(img).convert('RGB').copy()
>             r, w = 1, img.size[0]
>             for i in range(5):
>                 img = minmax_size(input_image.resize((w, int(
>                     input_image.size[1]*r)), Image.BILINEAR if r > 1 else Image.LANCZOS), args.max_dimensions, args.min_dimensions)
>                 t = test_transform(image=np.array(pad(img).convert('RGB')))['image'][:1].unsqueeze(0)
>                 w = image_resizer(t.to(args.device)).argmax(-1).item()*32
>                 if (w/img.size[0] == 1):
>                     break
>                 r *= w/img.size[0]
>         return t.to(args.device)
>     else:
>         return test_transform(image=np.array(pad(img).convert('RGB')))['image'][:1].unsqueeze(0)
> 
> 
> image_tensors = [img_resize(pil) for pil in pils]
> shapes = np.array([list(t.shape) for t in image_tensors])
> 
> batches = []
> indices = []
> shape_nums = (shapes[:, [-2, -1]]//32*np.array([args.max_width//32, 1])).sum(1)
> for shape in np.unique(shapes, axis=0):
>     inds = np.where(shape_nums == (args.max_width//32*shape[2]+shape[3])//32)[0]
>     #print(shape, inds, shape_nums[inds])
>     indices.extend(list(inds))
>     tensors = torch.cat([image_tensors[j] for j in inds])
>     for i in range(0, len(tensors), batchsize):
>         batches.append(tensors[i:i+batchsize])
> 
> 
> outputs = []
> for batch in tqdm(batches):
>     outputs.extend(predict(batch))
> outputs = [outputs[i] for i in np.argsort(indices)]
> 
> print('\n'.join(outputs))
> ```

I will try to implement with base on this code. Thank you so much

**Comment by cnxdeverloper** [comment]
> Yes, that is an issue here. While you can pad the images of one batch to the same size, the performance will most likely suffer and it would probably take even longer.
> So you will only be able to batch images of similar size. I tried it out, it does speed up the prediction by a little bit:
> I've batched together all images of the same size after the resizing step (I think this one has to be done indiviudally, not sure how to handle different image sizes here)
> 
> ```python
> # LaTeX-OCR/batchprediction.py
> import os
> import torch
> import numpy as np
> from dataset.dataset import *
> from pix2tex import *
> from PIL import Image
> from tqdm.auto import tqdm
> from munch import Munch
> 
> img_path = ...
> imgs = os.listdir(img_path)
> pils = [Image.open(os.path.join(img_path, p)) for p in imgs]
> batchsize = 5
> 
> 
> args = Munch({'config': 'settings/config.yaml', 'checkpoint': 'checkpoints/weights.pth', 'no_cuda': False, 'no_resize': False})
> 
> args, model, image_resizer, tokenizer = initialize(args)
> encoder, decoder = model.encoder, model.decoder
> 
> 
> def predict(im):
>     with torch.no_grad():
>         model.eval()
>         device = args.device
>         encoded = encoder(im.to(device))
>         dec = model.decoder.generate(torch.LongTensor([args.bos_token]*len(encoded))[:, None].to(device), args.max_seq_len,
>                                      eos_token=args.pad_token, context=encoded, temperature=args.get('temperature', .2))
>         pred = token2str(dec, tokenizer)
>     return [post_process(s) for s in pred]
> 
> 
> def img_resize(img):
>     img = minmax_size(pad(img), args.max_dimensions, args.min_dimensions)
>     if image_resizer is not None and not args.no_resize:
>         with torch.no_grad():
>             input_image = pad(img).convert('RGB').copy()
>             r, w = 1, img.size[0]
>             for i in range(5):
>                 img = minmax_size(input_image.resize((w, int(
>                     input_image.size[1]*r)), Image.BILINEAR if r > 1 else Image.LANCZOS), args.max_dimensions, args.min_dimensions)
>                 t = test_transform(image=np.array(pad(img).convert('RGB')))['image'][:1].unsqueeze(0)
>                 w = image_resizer(t.to(args.device)).argmax(-1).item()*32
>                 if (w/img.size[0] == 1):
>                     break
>                 r *= w/img.size[0]
>         return t.to(args.device)
>     else:
>         return test_transform(image=np.array(pad(img).convert('RGB')))['image'][:1].unsqueeze(0)
> 
> 
> image_tensors = [img_resize(pil) for pil in pils]
> shapes = np.array([list(t.shape) for t in image_tensors])
> 
> batches = []
> indices = []
> shape_nums = (shapes[:, [-2, -1]]//32*np.array([args.max_width//32, 1])).sum(1)
> for shape in np.unique(shapes, axis=0):
>     inds = np.where(shape_nums == (args.max_width//32*shape[2]+shape[3])//32)[0]
>     #print(shape, inds, shape_nums[inds])
>     indices.extend(list(inds))
>     tensors = torch.cat([image_tensors[j] for j in inds])
>     for i in range(0, len(tensors), batchsize):
>         batches.append(tensors[i:i+batchsize])
> 
> 
> outputs = []
> for batch in tqdm(batches):
>     outputs.extend(predict(batch))
> outputs = [outputs[i] for i in np.argsort(indices)]
> 
> print('\n'.join(outputs))
> ```

I will try to implement with base on this code. Thank you so much

**Comment by cnxdeverloper** [comment]
Dear author.
I applied your code run with batch_size and It worked and speed up a little bit performance. I saw the decoder.generate() was very time consuming and I saw u applied for _ in range(seq_len): to call model inference in this function. Do you have any idealto improve it

**Comment by lukas-blecher** [comment]
There is only little to be optimized in the prediction call.
`for _ in range(seq_len)` is necessary because that's where the autoregressive sampling is happening. It's not possible to parallelize this part if you wanted to do that.
I don't know where much more speed up should come from. I guess the model is a little bit faster once it is traced with ONNX or similar #2 but I haven't found a way to do that yet.
